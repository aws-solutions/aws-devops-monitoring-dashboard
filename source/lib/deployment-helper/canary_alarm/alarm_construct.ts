// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Construct } from 'constructs';
import * as cw from 'aws-cdk-lib/aws-cloudwatch';
import { addCfnSuppressRules } from '@aws-solutions-constructs/core';

export interface AlarmProps {
  readonly canaryName: string;
  readonly evalPeriods: number;
  readonly alarmPeriods: number;
  readonly threshold: number;
  alarmName?: string;
  index?: number;
}

export class AlarmConstruct extends Construct {
  constructor(scope: Construct, id: string, props: AlarmProps) {
    super(scope, id);

    if (props.index === undefined) {
      props.index = 1;
    }
    if (props.alarmName === undefined) {
      props.alarmName = 'Synthetics-Alarm-' + props.canaryName + '-' + props.index;
    }
    // Alarm State - the canary is "failed" when...
    // --------------------------------------------
    // If you choose to create alarms, they are created with the following
    // name convention:Synthetics-Alarm-canaryName-index
    const alarm = new cw.Alarm(this, 'AppAlarm', {
      alarmName: props.alarmName,
      metric: new cw.Metric({
        namespace: 'CloudWatchSynthetics',
        metricName: 'SuccessPercent',
        dimensionsMap: { CanaryName: props.canaryName }
      }),
      threshold: props.threshold,
      comparisonOperator: cw.ComparisonOperator.LESS_THAN_THRESHOLD,
      evaluationPeriods: props.evalPeriods,
      datapointsToAlarm: props.alarmPeriods,
      alarmDescription: 'Alarm when canary success is less than 100% on the most recent check.'
    });

    const alarmResource = alarm.node.findChild('Resource') as cw.CfnAlarm;
    addCfnSuppressRules(alarmResource, [
      {
        id: 'W28',
        reason: 'Static names chosen intentionally to provide fixed name structure required in the solution'
      }
    ]);
  }
}
