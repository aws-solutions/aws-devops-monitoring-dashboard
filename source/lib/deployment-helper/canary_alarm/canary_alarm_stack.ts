// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * @description
 * Creates a Canary that monitors an http or https url, and an alarm that will
 * send a signal to the solution each time the app is not 100% available.
 * Optionally, an S3 bucket for canary artifacts and access logging bucket are
 * created.
 * @author @aws-solutions
 */

import { Aws, CfnCondition, CfnOutput, CfnParameter, Fn, RemovalPolicy, Stack, StackProps, Tags } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { CfnCanary } from 'aws-cdk-lib/aws-synthetics';
import { Canary, Code, Runtime, Schedule, Test } from '@aws-cdk/aws-synthetics-alpha';
import {
  BlockPublicAccess,
  Bucket,
  BucketAccessControl,
  BucketEncryption,
  BucketProps,
  CfnBucket,
  CfnBucketPolicy,
  IBucket
} from 'aws-cdk-lib/aws-s3';
import { AnyPrincipal, CfnRole, Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { AlarmConstruct } from './alarm_construct';
import { addCfnSuppressRules } from '@aws-solutions-constructs/core';
import { NagSuppressions } from 'cdk-nag';

export interface CanaryProps extends StackProps {
  readonly solutionId: string;
  readonly solutionVersion?: string;
  readonly solutionDistBucket?: string;
  readonly solutionDistName?: string;
  readonly solutionName: string;
}

export class CanaryStack extends Stack {
  /**
   * @function Object() { [native code] }
   * @param {Construct} scope - parent of the construct
   * @param props
   * @param {string} id - identifier for the object
   */
  constructor(scope: Construct, id: string, props: CanaryProps) {
    super(scope, id, props);
    //=========================================================================
    // PARAMETERS
    //=========================================================================
    /**
     * @description name of the CodeCommit repository for the application to be
     * monitored
     * @type {CfnParameter}
     */
    const paramRepoName = new CfnParameter(this, 'RepoName', {
      description: 'Name of CodeCommit repository for the application.',
      type: 'String'
    });

    /**
     * @description name of the application
     * @type {CfnParameter}
     */
    const paramAppName = new CfnParameter(this, 'AppName', {
      description: 'Name of the application that canary monitors.',
      type: 'String'
    });

    /**
     * @description url to monitor via a synthetic transaction
     * @type {CfnParameter}
     */
    const paramCanaryUrl = new CfnParameter(this, 'URL', {
      description:
        'Application or endpoint URL you want to monitor with the canary (for example, https://www.example.com). The canary will check the site every 5 minutes.',
      type: 'String'
    });

    /**
     * @description name of the canary (synthetic transaction)
     * @type {CfnParameter}
     */
    const paramCanaryName = new CfnParameter(this, 'CanaryName', {
      description: 'Name of your Canary (new or existing) so users can easily understand what it is in the console.',
      type: 'String',
      default: 'mycanary',
      allowedPattern: '^[0-9a-z_-]+$'
    });

    /**
     * @description milliseconds to wait for a response
     * @type {CfnParameter}
     */
    const paramResponseThresh = new CfnParameter(this, 'ResponseThreshold', {
      description: 'Number of milliseconds to wait for a url response before considering the canary failed.',
      type: 'String',
      default: '15000',
      allowedPattern: '^[0-9]+$'
    });

    /**
     * @description canary interval, in minutes
     * @type {CfnParameter}
     */
    const paramCanaryInterval = new CfnParameter(this, 'Interval', {
      description: 'Interval, in minutes.',
      type: 'Number',
      default: 5
    });

    /**
     * @description threshold for SuccessPercent
     * @type {CfnParameter}
     */
    const paramThreshold = new CfnParameter(this, 'PercentThreshold', {
      description:
        'Threshold for Success (percentage). Any value less than this value will result in an alarm being triggered.',
      type: 'Number',
      default: 100
    });

    /**
     * @description number of eval periods to compare to threshold
     * @type {CfnParameter}
     */
    const paramEvalPeriods = new CfnParameter(this, 'EvalPeriods', {
      description: 'Number of periods to compare to the threshold.',
      type: 'Number',
      default: 1
    });

    /**
     * @description number of periods before alarming when below threshold
     * @type {CfnParameter}
     */
    const paramAlarmPeriods = new CfnParameter(this, 'AlarmPeriods', {
      description:
        'Number of collection periods over which the threshold is exceeded before alarming. ' +
        'This value must be less than or equal to Evaluation Periods',
      type: 'Number',
      default: 1
    });

    /**
     * @description should artifact bucket be created?
     * @type {CfnParameter}
     */
    const paramCreateBucket = new CfnParameter(this, 'CreateBucket', {
      description:
        'Canaries store artifacts in an S3 bucket. ' +
        'Should this canary create a new bucket? Enter the bucket name (new or existing) below.',
      type: 'String',
      default: 'No',
      allowedValues: ['No', 'Yes']
    });

    /**
     * @description should canary be created?
     * @type {CfnParameter}
     */
    const paramCreateCanary = new CfnParameter(this, 'CreateCanary', {
      description:
        'Should a new canary be created? If yes, a canary with the name specified above will be created. If no, just skip the rest of Canary Configuration and move on to Application Monitoring.',
      type: 'String',
      default: 'Yes',
      allowedValues: ['No', 'Yes']
    });

    /**
     * @description bucket to store canary artifacts. Can be an existing bucket.
     * @type {CfnParameter}
     */
    const paramBucketName = new CfnParameter(this, 'BucketName', {
      description:
        'Name of the bucket (new or existing) for logging Canary artifacts. ' +
        'Each Canary will log to this bucket to a different prefix.',
      type: 'String'
    });

    const parameterMetaData = {
      'AWS::CloudFormation::Interface': {
        ParameterGroups: [
          {
            Label: {
              default: 'Canary Configuration'
            },
            Parameters: [
              'CanaryName',
              'CreateCanary',
              'URL',
              'Interval',
              'ResponseThreshold',
              'CreateBucket',
              'BucketName'
            ]
          },
          {
            Label: {
              default: 'Application Monitoring'
            },
            Parameters: ['AppName', 'RepoName']
          },
          {
            Label: {
              default: 'Alarm Configuration'
            },
            Parameters: ['PercentThreshold', 'EvalPeriods', 'AlarmPeriods']
          }
        ],
        ParameterLabels: {
          URL: {
            default: 'URL'
          },
          CanaryName: {
            default: 'Canary Name'
          },
          CreateCanary: {
            default: 'Create New Canary?'
          },
          CanaryInterval: {
            default: 'Canary Interval'
          },
          CreateBucket: {
            default: 'Create Artifact Bucket?'
          },
          BucketName: {
            default: 'Artifact Bucket Name'
          },
          PercentThreshold: {
            default: '% Success Threshold (<)'
          },
          ResponseThreshold: {
            default: 'Response Threshold (ms)'
          },
          EvalPeriods: {
            default: 'Evaluation Periods'
          },
          AlarmPeriods: {
            default: 'Alarm Periods'
          },
          AppName: {
            default: 'Application Name'
          },
          RepoName: {
            default: 'Repository Name'
          }
        }
      }
    };
    this.templateOptions.metadata = parameterMetaData;

    //=========================================================================
    // CONDITIONS
    //=========================================================================
    const cndCanaryBucket = new CfnCondition(this, 'CanaryBucketCondition', {
      expression: Fn.conditionEquals(paramCreateBucket, 'Yes')
    });

    const cndCreateCanary = new CfnCondition(this, 'CreateCanaryCondition', {
      expression: Fn.conditionEquals(paramCreateCanary, 'Yes')
    });

    const cndCreateBucket = new CfnCondition(this, 'CreateBucketCondition', {
      expression: Fn.conditionAnd(cndCanaryBucket, cndCreateCanary)
    });

    //=========================================================================
    // RESOURCES
    //=========================================================================
    // ArtifactBucket
    // --------------
    const ArtifactBucketCfg: BucketProps = {
      encryption: BucketEncryption.S3_MANAGED,
      versioned: false,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      removalPolicy: RemovalPolicy.RETAIN,
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      bucketName: paramBucketName.valueAsString
    };

    const bktArtifact = new Bucket(this, 'ArtifactBucket', ArtifactBucketCfg);

    bktArtifact.addToResourcePolicy(
      new PolicyStatement({
        sid: 'HttpsOnly',
        resources: [bktArtifact.bucketArn, bktArtifact.arnForObjects('*')],
        actions: ['s3:*'],
        principals: [new AnyPrincipal()],
        effect: Effect.DENY,
        conditions: {
          Bool: {
            'aws:SecureTransport': 'false'
          }
        }
      })
    );

    Tags.of(bktArtifact).add('Name', props.solutionName + ' Canary Artifacts');

    const cfnArtifactBucket = <CfnBucket>bktArtifact.node.defaultChild;
    cfnArtifactBucket.addPropertyDeletionOverride('LifecycleConfiguration.Rules');
    cfnArtifactBucket.cfnOptions.condition = cndCreateBucket;

    const cfnBucketPolicy = <CfnBucketPolicy>bktArtifact.node.findChild('Policy').node.defaultChild;
    cfnBucketPolicy.cfnOptions.condition = cndCreateBucket;

    const bktArtifactResource = bktArtifact.node.findChild('Resource') as CfnBucket;
    addCfnSuppressRules(bktArtifactResource, [
      {
        id: 'W35',
        reason: 'This bucket is used by canary to store artifacts and no access logging bucket is needed.'
      }
    ]);

    // Add cdk-nag suppression
    NagSuppressions.addResourceSuppressions(bktArtifactResource, [
      {
        id: 'AwsSolutions-S1',
        reason: 'This bucket is used by canary to store artifacts and no access logging bucket is needed.'
      }
    ]);

    // Get iBucket reference - works whether bucket was created or not
    //
    const artifactBucket: IBucket = Bucket.fromBucketAttributes(this, 'ArtifactsBucket', {
      bucketName: paramBucketName.valueAsString
    });

    const canaryCode = [
      "var synthetics = require('Synthetics');",
      "const log = require('SyntheticsLogger');",
      '',
      'const pageLoadBlueprint = async function () {',
      '',
      '  const URL = "' + paramCanaryUrl.valueAsString + '";',
      '',
      '  let page = await synthetics.getPage();',
      "  const response = await page.goto(URL, {waitUntil: 'domcontentloaded', timeout: 30000});",
      '  //Wait for page to render.',
      '  //Increase or decrease wait time based on endpoint being monitored.',
      '  await page.waitFor(' + paramResponseThresh.valueAsString + ');',
      '  // This will take a screenshot that will be included in test output artifacts',
      "  await synthetics.takeScreenshot('loaded', 'loaded');",
      '  let pageTitle = await page.title();',
      "  log.info('Page title: ' + pageTitle);",
      '  if (response.status() !== 200) {',
      '      throw "Failed to load page!";',
      '  }',
      '};',
      '',
      'exports.handler = async () => {',
      '    return await pageLoadBlueprint();',
      '};'
    ];

    const intervalCondition = new CfnCondition(this, 'IntervalEqualsOne', {
      expression: Fn.conditionEquals(paramCanaryInterval.valueAsNumber, 1)
    });
    const scheduleExpression = Fn.conditionIf(
      intervalCondition.logicalId,
      'rate(1 minute)',
      `rate(${Fn.ref(paramCanaryInterval.logicalId)} minutes)`
    );

    const canary = new Canary(this, 'HttpCanary', {
      canaryName: paramCanaryName.valueAsString,
      schedule: Schedule.expression(scheduleExpression.toString()),
      test: Test.custom({
        code: Code.fromInline(canaryCode.join('\n')),
        handler: 'index.handler'
      }),
      artifactsBucketLocation: {
        bucket: artifactBucket,
        prefix: Aws.STACK_NAME
      },
      runtime: Runtime.SYNTHETICS_NODEJS_PUPPETEER_3_8
    });

    const canaryServiceRoleResource = canary.node.findChild('ServiceRole').node.findChild('Resource') as CfnRole;
    addCfnSuppressRules(canaryServiceRoleResource, [
      {
        id: 'W11',
        reason: 'Resource * is required by the canary service role.'
      }
    ]);
    // Add cdk-nag suppression
    NagSuppressions.addResourceSuppressions(canaryServiceRoleResource, [
      {
        id: 'AwsSolutions-IAM5',
        reason: 'Resource * is required by the canary service role.'
      }
    ]);

    const refCanary = canary.node.defaultChild as CfnCanary;
    refCanary.cfnOptions.condition = cndCreateCanary;

    // Alarm State - the canary is "failed" when...
    // --------------------------------------------
    const alarmName =
      props.solutionId + '-[' + paramAppName.valueAsString + ']-[' + paramRepoName.valueAsString + ']-MTTR';
    new AlarmConstruct(this, 'Alarm', {
      canaryName: paramCanaryName.valueAsString,
      alarmName: alarmName,
      evalPeriods: paramEvalPeriods.valueAsNumber,
      alarmPeriods: paramAlarmPeriods.valueAsNumber,
      threshold: paramThreshold.valueAsNumber
    });

    //=========================================================================
    // OUTPUTS
    //=========================================================================
    new CfnOutput(this, 'SolutionVersion', {
      value: props.solutionVersion ? props.solutionVersion : '',
      description: 'Version for DevOps Monitoring Dashboard on AWS Solution'
    });

    new CfnOutput(this, 'Canary Name', {
      value: paramCanaryName.valueAsString,
      description: 'Name of the Canary'
    });

    new CfnOutput(this, 'Alarm Name', {
      value: alarmName,
      description: 'Name of the Canary Alarm'
    });
  }
}
