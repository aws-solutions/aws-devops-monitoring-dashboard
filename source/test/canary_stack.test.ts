// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as cdk from 'aws-cdk-lib';
import { CanaryStack } from '../lib/deployment-helper/canary_alarm/canary_alarm_stack';
import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';

const SOLUTION_ID = 'SO0143';
const SOLUTION_NAME = 'DevOps Monitoring Dashboard on AWS';
const DIST_VERSION = 'v1.0.0';
const DIST_OUTPUT_BUCKET = 'devops_dashboard_test_bucket';
const DIST_SOLUTION_NAME = 'aws_devops_monitoring_dashboard';

/*
 * Snapshot test for Canary Alarm.
 */
test('Snapshot test for canary alarm', () => {
  const app = new cdk.App();
  const stack = new CanaryStack(app, 'CanaryAlarm', {
    solutionId: SOLUTION_ID,
    solutionVersion: DIST_VERSION,
    solutionName: SOLUTION_NAME,
    solutionDistBucket: DIST_OUTPUT_BUCKET,
    solutionDistName: DIST_SOLUTION_NAME
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
