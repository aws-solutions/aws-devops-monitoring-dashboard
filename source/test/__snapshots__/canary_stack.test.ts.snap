// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for canary alarm 1`] = `
{
  "Conditions": {
    "CanaryBucketCondition": {
      "Fn::Equals": [
        {
          "Ref": "CreateBucket",
        },
        "Yes",
      ],
    },
    "CreateBucketCondition": {
      "Fn::And": [
        {
          "Condition": "CanaryBucketCondition",
        },
        {
          "Condition": "CreateCanaryCondition",
        },
      ],
    },
    "CreateCanaryCondition": {
      "Fn::Equals": [
        {
          "Ref": "CreateCanary",
        },
        "Yes",
      ],
    },
    "IntervalEqualsOne": {
      "Fn::Equals": [
        {
          "Ref": "Interval",
        },
        1,
      ],
    },
  },
  "Metadata": {
    "AWS::CloudFormation::Interface": {
      "ParameterGroups": [
        {
          "Label": {
            "default": "Canary Configuration",
          },
          "Parameters": [
            "CanaryName",
            "CreateCanary",
            "URL",
            "Interval",
            "ResponseThreshold",
            "CreateBucket",
            "BucketName",
          ],
        },
        {
          "Label": {
            "default": "Application Monitoring",
          },
          "Parameters": [
            "AppName",
            "RepoName",
          ],
        },
        {
          "Label": {
            "default": "Alarm Configuration",
          },
          "Parameters": [
            "PercentThreshold",
            "EvalPeriods",
            "AlarmPeriods",
          ],
        },
      ],
      "ParameterLabels": {
        "AlarmPeriods": {
          "default": "Alarm Periods",
        },
        "AppName": {
          "default": "Application Name",
        },
        "BucketName": {
          "default": "Artifact Bucket Name",
        },
        "CanaryInterval": {
          "default": "Canary Interval",
        },
        "CanaryName": {
          "default": "Canary Name",
        },
        "CreateBucket": {
          "default": "Create Artifact Bucket?",
        },
        "CreateCanary": {
          "default": "Create New Canary?",
        },
        "EvalPeriods": {
          "default": "Evaluation Periods",
        },
        "PercentThreshold": {
          "default": "% Success Threshold (<)",
        },
        "RepoName": {
          "default": "Repository Name",
        },
        "ResponseThreshold": {
          "default": "Response Threshold (ms)",
        },
        "URL": {
          "default": "URL",
        },
      },
    },
  },
  "Outputs": {
    "AlarmName": {
      "Description": "Name of the Canary Alarm",
      "Value": {
        "Fn::Join": [
          "",
          [
            "SO0143-[",
            {
              "Ref": "AppName",
            },
            "]-[",
            {
              "Ref": "RepoName",
            },
            "]-MTTR",
          ],
        ],
      },
    },
    "CanaryName": {
      "Description": "Name of the Canary",
      "Value": {
        "Ref": "CanaryName",
      },
    },
    "SolutionVersion": {
      "Description": "Version for DevOps Monitoring Dashboard on AWS Solution",
      "Value": "v1.0.0",
    },
  },
  "Parameters": {
    "AlarmPeriods": {
      "Default": 1,
      "Description": "Number of collection periods over which the threshold is exceeded before alarming. This value must be less than or equal to Evaluation Periods",
      "Type": "Number",
    },
    "AppName": {
      "Description": "Name of the application that canary monitors.",
      "Type": "String",
    },
    "BucketName": {
      "Description": "Name of the bucket (new or existing) for logging Canary artifacts. Each Canary will log to this bucket to a different prefix.",
      "Type": "String",
    },
    "CanaryName": {
      "AllowedPattern": "^[0-9a-z_-]+$",
      "Default": "mycanary",
      "Description": "Name of your Canary (new or existing) so users can easily understand what it is in the console.",
      "Type": "String",
    },
    "CreateBucket": {
      "AllowedValues": [
        "No",
        "Yes",
      ],
      "Default": "No",
      "Description": "Canaries store artifacts in an S3 bucket. Should this canary create a new bucket? Enter the bucket name (new or existing) below.",
      "Type": "String",
    },
    "CreateCanary": {
      "AllowedValues": [
        "No",
        "Yes",
      ],
      "Default": "Yes",
      "Description": "Should a new canary be created? If yes, a canary with the name specified above will be created. If no, just skip the rest of Canary Configuration and move on to Application Monitoring.",
      "Type": "String",
    },
    "EvalPeriods": {
      "Default": 1,
      "Description": "Number of periods to compare to the threshold.",
      "Type": "Number",
    },
    "Interval": {
      "Default": 5,
      "Description": "Interval, in minutes.",
      "Type": "Number",
    },
    "PercentThreshold": {
      "Default": 100,
      "Description": "Threshold for Success (percentage). Any value less than this value will result in an alarm being triggered.",
      "Type": "Number",
    },
    "RepoName": {
      "Description": "Name of CodeCommit repository for the application.",
      "Type": "String",
    },
    "ResponseThreshold": {
      "AllowedPattern": "^[0-9]+$",
      "Default": "15000",
      "Description": "Number of milliseconds to wait for a url response before considering the canary failed.",
      "Type": "String",
    },
    "URL": {
      "Description": "Application or endpoint URL you want to monitor with the canary (for example, https://www.example.com). The canary will check the site every 5 minutes.",
      "Type": "String",
    },
  },
  "Resources": {
    "AlarmAppAlarm83138885": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W28",
              "reason": "Static names chosen intentionally to provide fixed name structure required in the solution",
            },
          ],
        },
      },
      "Properties": {
        "AlarmDescription": "Alarm when canary success is less than 100% on the most recent check.",
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "SO0143-[",
              {
                "Ref": "AppName",
              },
              "]-[",
              {
                "Ref": "RepoName",
              },
              "]-MTTR",
            ],
          ],
        },
        "ComparisonOperator": "LessThanThreshold",
        "DatapointsToAlarm": {
          "Ref": "AlarmPeriods",
        },
        "Dimensions": [
          {
            "Name": "CanaryName",
            "Value": {
              "Ref": "CanaryName",
            },
          },
        ],
        "EvaluationPeriods": {
          "Ref": "EvalPeriods",
        },
        "MetricName": "SuccessPercent",
        "Namespace": "CloudWatchSynthetics",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": {
          "Ref": "PercentThreshold",
        },
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ArtifactBucket7410C9EF": {
      "Condition": "CreateBucketCondition",
      "DeletionPolicy": "Retain",
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-S1",
              "reason": "This bucket is used by canary to store artifacts and no access logging bucket is needed.",
            },
          ],
        },
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W35",
              "reason": "This bucket is used by canary to store artifacts and no access logging bucket is needed.",
            },
          ],
        },
      },
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": {
          "Ref": "BucketName",
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "DevOps Monitoring Dashboard on AWS Canary Artifacts",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "ArtifactBucketPolicy4B4B7752": {
      "Condition": "CreateBucketCondition",
      "Properties": {
        "Bucket": {
          "Ref": "ArtifactBucket7410C9EF",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "ArtifactBucket7410C9EF",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "ArtifactBucket7410C9EF",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
              "Sid": "HttpsOnly",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "HttpCanaryE06DE534": {
      "Condition": "CreateCanaryCondition",
      "Properties": {
        "ArtifactS3Location": {
          "Fn::Join": [
            "",
            [
              "s3://",
              {
                "Ref": "BucketName",
              },
              "/",
              {
                "Ref": "AWS::StackName",
              },
            ],
          ],
        },
        "Code": {
          "Handler": "index.handler",
          "Script": {
            "Fn::Join": [
              "",
              [
                "var synthetics = require('Synthetics');
const log = require('SyntheticsLogger');

const pageLoadBlueprint = async function () {

  const URL = "",
                {
                  "Ref": "URL",
                },
                "";

  let page = await synthetics.getPage();
  const response = await page.goto(URL, {waitUntil: 'domcontentloaded', timeout: 30000});
  //Wait for page to render.
  //Increase or decrease wait time based on endpoint being monitored.
  await page.waitForTimeout(",
                {
                  "Ref": "ResponseThreshold",
                },
                ");
  // This will take a screenshot that will be included in test output artifacts
  await synthetics.takeScreenshot('loaded', 'loaded');
  let pageTitle = await page.title();
  log.info('Page title: ' + pageTitle);
  if (response.status() !== 200) {
      throw "Failed to load page!";
  }
};

exports.handler = async () => {
    return await pageLoadBlueprint();
};",
              ],
            ],
          },
        },
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "HttpCanaryServiceRoleF71089F4",
            "Arn",
          ],
        },
        "Name": {
          "Ref": "CanaryName",
        },
        "RuntimeVersion": "syn-nodejs-puppeteer-6.2",
        "Schedule": {
          "DurationInSeconds": "0",
          "Expression": {
            "Fn::If": [
              "IntervalEqualsOne",
              "rate(1 minute)",
              {
                "Fn::Join": [
                  "",
                  [
                    "rate(",
                    {
                      "Ref": "Interval",
                    },
                    " minutes)",
                  ],
                ],
              },
            ],
          },
        },
        "StartCanaryAfterCreation": true,
      },
      "Type": "AWS::Synthetics::Canary",
    },
    "HttpCanaryServiceRoleF71089F4": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM5",
              "reason": "Resource * is required by the canary service role.",
            },
          ],
        },
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W11",
              "reason": "Resource * is required by the canary service role.",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:ListAllMyBuckets",
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": "s3:GetBucketLocation",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":s3:::",
                        {
                          "Ref": "BucketName",
                        },
                      ],
                    ],
                  },
                },
                {
                  "Action": "s3:PutObject",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":s3:::",
                        {
                          "Ref": "BucketName",
                        },
                        "/",
                        {
                          "Ref": "AWS::StackName",
                        },
                        "/*",
                      ],
                    ],
                  },
                },
                {
                  "Action": "cloudwatch:PutMetricData",
                  "Condition": {
                    "StringEquals": {
                      "cloudwatch:namespace": "CloudWatchSynthetics",
                    },
                  },
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": [
                    "logs:CreateLogStream",
                    "logs:CreateLogGroup",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/cwsyn-*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "canaryPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
